types/rlp_unmarshal.go

```diff
@@ -60,12 +80,29 @@ func (b *Block) UnmarshalRLPFrom(p *fastrlp.Parser, v *fastrlp.Value) error {
 		return err
 	}
 
-	for _, txn := range txns {
-		bTxn := &Transaction{}
-		if err := bTxn.UnmarshalRLPFrom(p, txn); err != nil {
+	for i := 0; i < len(txns); i++ {
+		// Non-legacy tx raw contains a tx type prefix in the beginning according to EIP-2718.
+		// Here we check if the first element is a tx type and unmarshal it first.
+		txType := LegacyTx
+		if txns[i].Type() == fastrlp.TypeBytes {
+			if err = txType.UnmarshalRLPFrom(p, txns[i]); err != nil {
+				return err
+			}
+
+			// Then we increment element number in order to go to the actual tx data raw below.
+			i++
+		}
+
+		bTxn := &Transaction{
+			Type: txType,
+		}
+
+		if err = bTxn.UnmarshalRLPFrom(p, txns[i]); err != nil {
 			return err
 		}
 
+		bTxn.ComputeHash()
+
 		b.Transactions = append(b.Transactions, bTxn)
 	}
```

Comment:
Should we extract this function to some helper one, as it is (almost) the same as the one in `rlp_unmarshal_storage.go` (https://github.com/0xPolygon/polygon-edge/pull/1114/files#diff-885da5543ea77aa838b1210b3eb06bf214b06f2f77e0c2134051e4628efcbb15R34)?

Only this line differs (maybe we could provide it as a callback argument)
`if err = bTxn.UnmarshalRLPFrom(p, txns[i]); err != nil {`

WDYT?

Reply:
Done